lib-ee/emqx_ee_bridge/src/gcp_pubsub/emqx_ee_bridge_gcp_pubsub_resource.erl

```diff
@@ -0,0 +1,599 @@
+    ?TRACE(
+        "QUERY_ASYNC",
+        "gcp_pubsub_received",
+        #{requests => Requests, connector => BridgeId, state => State}
+    ),
+    do_send_requests_async(State, Requests, ReplyFunAndArgs, BridgeId).
+
+-spec on_get_status(manager_id(), state()) -> connected | disconnected.
+on_get_status(InstanceId, State) ->
+    #{
+        connect_timeout := Timeout,
+        pool_name := PoolName
+    } = State,
+    case do_get_status(InstanceId, PoolName, Timeout) of
+        true ->
+            connected;
+        false ->
+            ?SLOG(error, #{
+                msg => "gcp_pubsub_bridge_get_status_failed",
+                state => State
+            }),
+            disconnected
+    end.
+
+%%-------------------------------------------------------------------------------------------------
+%% Helper fns
+%%-------------------------------------------------------------------------------------------------
+
+-spec ensure_jwt_worker(manager_id(), config()) ->
+    #{
+        jwt_worker_id := jwt_worker(),
+        project_id := binary()
+    }.
+ensure_jwt_worker(InstanceId, #{
+    service_account_json := ServiceAccountJSON,
+    pubsub_topic := PubSubTopic
+}) ->
+    #{
+        project_id := ProjectId,
+        private_key_id := KId,
+        private_key := PrivateKeyPEM,
+        client_email := ServiceAccountEmail
+    } = ServiceAccountJSON,
+    %% fixed for pubsub; trailing slash is important.
+    Aud = <<"https://pubsub.googleapis.com/">>,
+    ExpirationMS = timer:hours(1),
+    Alg = <<"RS256">>,
+    Config = #{
+        private_key => PrivateKeyPEM,
+        resource_id => InstanceId,
+        expiration => ExpirationMS,
+        table => ?JWT_TABLE,
+        iss => ServiceAccountEmail,
+        sub => ServiceAccountEmail,
+        aud => Aud,
+        kid => KId,
+        alg => Alg
+    },
+
+    JWTWorkerId = <<"gcp_pubsub_jwt_worker:", InstanceId/binary>>,
+    Worker =
+        case emqx_connector_jwt_sup:ensure_worker_present(JWTWorkerId, Config) of
+            {ok, Worker0} ->
+                Worker0;
+            Error ->
+                ?tp(
+                    gcp_pubsub_bridge_jwt_worker_failed_to_start,
+                    #{instance_id => InstanceId, reason => Error}
+                ),
+                ?SLOG(error, #{
+                    msg => "failed_to_start_gcp_pubsub_jwt_worker",
+                    instance_id => InstanceId,
+                    pubsub_topic => PubSubTopic,
+                    reason => Error
+                }),
+                _ = emqx_connector_jwt_sup:ensure_worker_deleted(JWTWorkerId),
+                throw(failed_to_start_jwt_worker)
+        end,
+    MRef = monitor(process, Worker),
+    Ref = emqx_connector_jwt_worker:ensure_jwt(Worker),
+
+    %% to ensure that this resource and its actions will be ready to
+    %% serve when started, we must ensure that the first JWT has been
+    %% produced by the worker.
+    receive
```

Comment:
Curious: from the code of `emqx_connector_jwt_worker`, it seems that token creation does not do anything blocking or related to external resources 🤔 why do we need timeout/monitor wrapping?

Also just an idea: maybe we generate the first token in `emqx_connector_jwt_sup:ensure_worker_present(JWTWorkerId, Config)` directly and then pass it to the renewing server.  So that we do not need to do any process-related things 🤔

Reply:
I tried to handle the credentials inside a process with a special `format_status` function to avoid accidentally logging credentials/keys when manipulating the keys.  That's also why I didn't attempt to produce the key in `init`, but only in `handle_continue` (errors in `init` might still log the credentials due to `start_link` erroring).